File: packages/core/src/game-skills/pygame/controller.py
# when_to_use: Gamepad / controller support for a Pygame game running
# on Pyodide. Wraps pygame.joystick with deadzones, named-button mapping
# (Xbox/PS standard layout), hot-plug detection, and a single `poll()`
# call your event loop reads each frame.
#
# Pyodide-specific note: pygame-ce on Pyodide bridges the browser
# Gamepad API through pygame.joystick. Connect events fire only AFTER
# the user presses a button on the controller (Chrome's policy). The
# first poll after that returns a Joystick object; before that, no pad
# is visible. Show "Press any button on your controller to begin"
# during onboarding.

import pygame

# Standard mapping (Xbox layout). PS players see ✕○□△ but the indices
# are the same on the Standard Gamepad mapping the browser exposes.
STANDARD_BUTTONS = [
    "a", "b", "x", "y",
    "lb", "rb",
    "lt", "rt",
    "select", "start",
    "l3", "r3",
    "up", "down", "left", "right",
    "home",
]


def _apply_deadzone(x: float, y: float, deadzone: float) -> tuple[float, float]:
    """Radial deadzone — same shape as the JS engine skills. Below the
    threshold returns (0, 0); above, smoothly remaps so the stick
    starts at 0 magnitude at the boundary."""
    mag = (x * x + y * y) ** 0.5
    if mag < deadzone:
        return (0.0, 0.0)
    scale = (mag - deadzone) / (1.0 - deadzone) / mag
    return (x * scale, y * scale)


class Controller:
    """One-pad wrapper. Multi-pad games instantiate two of these against
    different joystick indices; the user-facing brief usually means one."""

    def __init__(self, deadzone: float = 0.15):
        pygame.joystick.init()
        self.deadzone = deadzone
        self._joy: pygame.joystick.Joystick | None = None
        self._last_buttons: list[bool] = []
        self._just_pressed: set[str] = set()
        self._just_released: set[str] = set()

    def _ensure_joy(self) -> None:
        # Re-probe every poll — Pyodide's joystick adapter creates the
        # Joystick lazily on first browser-level button press. Cheap to
        # call (joystick count is cached).
        if self._joy is None and pygame.joystick.get_count() > 0:
            self._joy = pygame.joystick.Joystick(0)
            self._joy.init()
            self._last_buttons = [False] * self._joy.get_numbuttons()

    def poll(self) -> dict | None:
        """Return a snapshot of stick / button / trigger state, or None
        when no controller is connected. Updates internal edge-tracking
        (wasPressed / wasReleased)."""
        self._ensure_joy()
        if self._joy is None:
            return None
        joy = self._joy
        # Edge tracking
        for i in range(min(joy.get_numbuttons(), len(STANDARD_BUTTONS))):
            pressed_now = joy.get_button(i) == 1
            was_pressed = self._last_buttons[i] if i < len(self._last_buttons) else False
            name = STANDARD_BUTTONS[i] if i < len(STANDARD_BUTTONS) else f"b{i}"
            if pressed_now and not was_pressed:
                self._just_pressed.add(name)
            if not pressed_now and was_pressed:
                self._just_released.add(name)
            if i < len(self._last_buttons):
                self._last_buttons[i] = pressed_now

        # Standard Gamepad axis layout (browser-bridged):
        #   axis 0 = left stick X, axis 1 = left stick Y
        #   axis 2 = right stick X, axis 3 = right stick Y
        #   axes 4 / 5 are sometimes triggers; sometimes buttons[6/7].value.
        def _axis(i: int) -> float:
            return joy.get_axis(i) if joy.get_numaxes() > i else 0.0

        lx, ly = _apply_deadzone(_axis(0), _axis(1), self.deadzone)
        rx, ry = _apply_deadzone(_axis(2), _axis(3), self.deadzone)

        # D-pad on most pads is hat 0 — (-1, 0) = left, (0, 1) = up, etc.
        # Some browsers map it to buttons[12-15] instead; use whichever
        # the current pad exposes.
        dpad_x = dpad_y = 0
        if joy.get_numhats() > 0:
            dpad_x, dpad_y = joy.get_hat(0)

        return {
            "id": joy.get_name(),
            "buttons": {
                name: (joy.get_button(i) == 1)
                for i, name in enumerate(STANDARD_BUTTONS)
                if i < joy.get_numbuttons()
            },
            "leftStick": (lx, ly),
            "rightStick": (rx, ry),
            "dpad": (dpad_x, dpad_y),
            # Triggers vary across browsers — buttons[6/7].value isn't
            # exposed by pygame-ce. Approximate via axis 4/5 if present.
            "lt": max(0.0, _axis(4)),
            "rt": max(0.0, _axis(5)),
        }

    def was_pressed(self, name: str) -> bool:
        return name in self._just_pressed

    def was_released(self, name: str) -> bool:
        return name in self._just_released

    def is_connected(self) -> bool:
        return self._joy is not None

    def flush(self) -> None:
        """Call at the end of each frame after reading was_pressed / was_released."""
        self._just_pressed.clear()
        self._just_released.clear()

    def handle_event(self, event: pygame.event.Event) -> None:
        """Optionally route JOYDEVICEADDED / JOYDEVICEREMOVED events
        through here so the controller reconnects after a browser
        suspend/resume cycle. Most games can rely on _ensure_joy()
        instead."""
        if event.type == pygame.JOYDEVICEADDED:
            # New pad — re-probe on next poll
            self._joy = None
        elif event.type == pygame.JOYDEVICEREMOVED:
            self._joy = None
            self._last_buttons = []

    def dispose(self) -> None:
        if self._joy is not None:
            self._joy.quit()
            self._joy = None
        pygame.joystick.quit()


# Usage:
#   ctrl = Controller(deadzone=0.15)
#   while running:
#       for event in pygame.event.get():
#           if event.type == pygame.QUIT:
#               running = False
#           ctrl.handle_event(event)
#       state = ctrl.poll()
#       if state:
#           player.x += state["leftStick"][0] * speed * dt
#           player.y += state["leftStick"][1] * speed * dt
#           if ctrl.was_pressed("a"):
#               player.jump()
#           if state["buttons"]["rt"]:
#               player.fire()
#       ctrl.flush()
#       pygame.display.flip()
#       await asyncio.sleep(0)  # YIELD — Pyodide requires it (engine guide §A)

File: packages/core/src/game-skills/phaser/controller.js
// when_to_use: Gamepad / controller support for a Phaser scene. Wraps
// Phaser's input.gamepad plugin with deadzones, named-button mapping
// (Xbox/PS standard), connect/disconnect events, and an optional
// vibration helper. Call `setupController(this)` inside `create()`.
//
// IMPORTANT: gamepad input requires `input: { gamepad: true }` in your
// game config — Phaser disables it by default to avoid touching the
// browser API on first user gesture. Add it in the Phaser.Game config:
//
//   const game = new Phaser.Game({
//     ...,
//     input: { gamepad: true },
//   });

import Phaser from 'phaser';

const STANDARD_BUTTONS = [
  'a',
  'b',
  'x',
  'y',
  'lb',
  'rb',
  'lt',
  'rt',
  'select',
  'start',
  'l3',
  'r3',
  'up',
  'down',
  'left',
  'right',
  'home',
];

export function setupController(scene, { deadzone = 0.15 } = {}) {
  // Phaser's gamepad plugin lazily creates Pad objects on first connect.
  // We track the FIRST pad only (most game briefs need single-player
  // input); extending to multiple pads is `scene.input.gamepad.pad2/3/4`.
  let pad = scene.input.gamepad?.pad1 ?? null;
  const justPressed = new Set();
  const justReleased = new Set();
  const lastState = new Array(STANDARD_BUTTONS.length).fill(false);

  scene.input.gamepad?.once('connected', (newPad) => {
    pad = newPad;
    scene.events.emit('controller:connected', { id: newPad.id });
  });
  scene.input.gamepad?.on('disconnected', (gone) => {
    if (gone === pad) {
      pad = null;
      lastState.fill(false);
      scene.events.emit('controller:disconnected');
    }
  });

  // Radial deadzone — see Three.js controller skill for the rationale.
  // Phaser's pad.leftStick / rightStick already return Vector2 with
  // magnitude, but it doesn't apply a deadzone for you.
  function applyDeadzone(vec) {
    const mag = vec.length();
    if (mag < deadzone) return new Phaser.Math.Vector2(0, 0);
    const scale = (mag - deadzone) / (1 - deadzone) / mag;
    return new Phaser.Math.Vector2(vec.x * scale, vec.y * scale);
  }

  function poll() {
    if (pad === null || !pad.connected) return null;
    // Edge tracking. Phaser's pad.A / pad.B etc are convenience getters
    // that read .buttons[0].pressed; we reuse them via name lookup.
    for (let i = 0; i < pad.buttons.length && i < STANDARD_BUTTONS.length; i += 1) {
      const pressedNow = pad.buttons[i]?.pressed ?? false;
      const name = STANDARD_BUTTONS[i] ?? `b${i}`;
      if (pressedNow && !lastState[i]) justPressed.add(name);
      if (!pressedNow && lastState[i]) justReleased.add(name);
      lastState[i] = pressedNow;
    }
    const left = applyDeadzone(pad.leftStick);
    const right = applyDeadzone(pad.rightStick);
    return {
      id: pad.id,
      buttons: Object.fromEntries(
        STANDARD_BUTTONS.map((name, i) => [name, pad.buttons[i]?.pressed ?? false]),
      ),
      // Phaser exposes triggers as buttons[6/7].value (analog 0..1).
      lt: pad.buttons[6]?.value ?? 0,
      rt: pad.buttons[7]?.value ?? 0,
      leftStick: { x: left.x, y: left.y },
      rightStick: { x: right.x, y: right.y },
    };
  }

  function rumble(durationMs = 200, weakMagnitude = 0.5, strongMagnitude = 0.5) {
    // Phaser doesn't wrap the vibrationActuator API directly — fall
    // through to the underlying browser pad object via pad.pad.
    pad?.pad?.vibrationActuator?.playEffect?.('dual-rumble', {
      duration: durationMs,
      weakMagnitude,
      strongMagnitude,
    });
  }

  return {
    poll,
    rumble,
    isConnected: () => pad?.connected ?? false,
    wasPressed: (name) => justPressed.has(name),
    wasReleased: (name) => justReleased.has(name),
    /** Call at the end of each scene update() after reading wasPressed/wasReleased. */
    flush() {
      justPressed.clear();
      justReleased.clear();
    },
  };
}

// Usage:
//   class PlayScene extends Phaser.Scene {
//     create() {
//       this.controller = setupController(this);
//       this.events.on('controller:connected', (info) =>
//         console.log('pad connected:', info.id),
//       );
//     }
//     update(_t, dt) {
//       const state = this.controller.poll();
//       if (state) {
//         this.player.body.setVelocityX(state.leftStick.x * 200);
//         this.player.body.setVelocityY(state.leftStick.y * 200);
//         if (this.controller.wasPressed('a')) this.player.jump();
//         if (state.buttons.rt) this.fireWeapon();
//       }
//       this.controller.flush();
//     }
//     onHit() { this.controller.rumble(150, 0.4, 0.8); }
//   }
